### PR TITLE
HPCC-20619 Prevent backupnode and thor master are on the same node

### PR DESCRIPTION
**Pull request by xwang2713**
Prevent adding backupnode instance automatically.
Update validate environment function to check backupnode
Add backupnode function test.
Resume every component has at least one instance except backupnode if ip/ipfile provided

Signed-off-by: Xiaoming Wang <xiaoming.wang@lexisnexi.com> 

@kenrowland please help reviewing.
Here are the summary of the changes:
1) deployment/deployutils/wizardInputs.cpp:  
    Disable adding backupnode instance unless "-assign_ips backupnode <ip>" provided in envgen
2) deployment/deployutils/deployUtils.cpp:
    Add checking for backupnode and thor master on the same node
3) deployment/configenv/xml_jlibpt/EnvHelper.cpp, Software.cpp:
    Add BackupNode name resolution and instance generation
4) deployment/configenv/xml_jlibpt/SWBackupNode.hpp/cpp:
    Overwrite add method to handle adding (multiple) <NodeGroup/> element
5) deployment/envgen2/test/test.sh add-node 
    Add a test case for backupnode. 
6)  deployment/envgen2/test/src/result-comparison/*.diff
     These file are for test results comparision since some entries in environment.xml changed in rc2.
     These can be ignored in reviewing procedure.
7) deployment/envgen/main.cpp
    This is not related to backupnode. It just ensure to create 1 thor and 1 roxie if users do not provide "-thornodes" and "-roxienodes". 

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [ ] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Testing:

Build: http://10.240.32.243/view/custom/job/CE-Platform-JIRA-HPCC-20619/
Tested with envgen:
    envgen  -env /etc/HPCCSystems/environment.xml -ip .
Tested with envgen2 in a cluster which has 1 support node, 2 thor and 2 roxie

### CODE REVIEW

**Review by kenrowland — COMMENTED** [review]
Only one comment, I think. Otherwise, looks reasonable.
@xwang2713

File: deployment/configenv/xml_jlibpt/SWBackupNode.cpp [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+   const char* selector = params->queryProp("@selector");
+   if (selector && !stricmp("NodeGroup", selector))
+   {
+       IPropertyTree *nodeGroup = createPTree(selector);
+       IPropertyTree* pAttrs = params->queryPropTree("Attributes");
+       updateNode(nodeGroup, pAttrs, NULL);
+       if (!nodeGroup->hasProp("@interval"))
+       {
+          xpath.clear().appendf("xs:element/xs:complexType/xs:sequence/xs:element[@name=\"NodeGroup\"]/xs:complexType/xs:attribute[@name=\"interval\"]/@default");
+          const char *interval = m_pSchema->queryProp(xpath.str());
```
[the remark this answers is not in the document]
Comment (xwang2713) [reply deployment/configenv/xml_jlibpt/SWBackupNode.cpp]: If xsd file changed I need change default value handling.

**Comment by richardkchapman** [comment]
@kenrowland Are you happy with this now?

**Review by kenrowland — APPROVED** [review]
Looks good

**Pull request merged**